lms/services/application_instance.py

```diff
@@ -35,6 +36,85 @@ def __init__(self, application_instance: ApplicationInstance):
         )
 
 
+@dataclass
+class SearchBuilder:
+    base_model: BASE
+    clause_builders: dict
+    key_map: dict = field(default_factory=lambda: {"id_": "id", "type_": "type"})
+    query_start: Optional[callable] = None
+
+    def get_query(self, db_session, kwargs, combination=sa.and_):
+        query = self.query_start(db_session) or db_session.query(self.base_model)
+
+        return self.filter_query(query, kwargs, combination=combination)
+
+    def filter_query(self, query, kwargs, combination=sa.and_):
+        if clauses := list(self.get_clauses(kwargs)):
+            query = query.filter(combination(*clauses))
```

Comment:
Because we deal with everything is discrete clauses, we can choose how to combine them (AND / OR) easily.